src/analysis/complex/upper_half_plane.lean

```diff
@@ -0,0 +1,169 @@
+/-
+Copyright (c) 2021 Alex Kontorovich and Heather Macbeth and Marc Masdeu. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Alex Kontorovich, Heather Macbeth, Marc Masdeu
+-/
+
+import linear_algebra.special_linear_group
+import analysis.complex.basic
+import group_theory.group_action.defs
+
+/-!
+# The upper half plane and its automorphisms
+
+This file defines `upper_half_plane` to be the upper half plane in `ℂ`.
+
+We furthermore equip it with the structure of an `SL(2,ℝ)` action by
+fractional linear transformations.
+
+We define the notation `ℍ` for the upper half plane available in the locale
+`upper_half_plane` so as not to conflict with the quaternions.
+-/
+
+noncomputable theory
+
+open matrix matrix.special_linear_group
+
+open_locale classical big_operators
+
+local attribute [instance] fintype.card_fin_even
+
+/-- The upper half plane -/
+abbreviation upper_half_plane :=
+{point : ℂ // 0 < point.im}
+
+localized "notation `ℍ` := upper_half_plane" in upper_half_plane
+
+namespace upper_half_plane
+
+/-- Imaginary part -/
+def im (z : ℍ) := z.1.im
+
+/-- Real part -/
+def re (z : ℍ) := z.1.re
```

Comment:
I think it would make sense to define these instead as:
```suggestion
/-- Imaginary part -/
def im (z : ℍ) := (z : ℂ).im

/-- Real part -/
def re (z : ℍ) := (z : ℂ).re
```
which unfold more nicely. Nothing below would need to change.